doc/man7/provider-rand.pod

```diff
@@ -0,0 +1,275 @@
+algorithms and random number sources and make
+them available to applications via the API function L<EVP_RAND(3)>.
+
+=head2 Context Management Functions
+
+OP_rand_newctx() should create and return a pointer to a provider side
+structure for holding context information during a rand operation.
+A pointer to this context will be passed back in a number of the other rand
+operation function calls.
+The parameter I<provctx> is the provider context generated during provider
+initialisation (see L<provider(7)>).
+The parameter I<parent> specifies an other rand instance to be used for
+seeding purposes.  If NULL, the operating system will be used for seeding.
+The parameter I<parent_calls> points to dispatch table for I<parent>.
+Thus, the parent need not be from the same provider as the new instance.
+
+OP_rand_freectx() is passed a pointer to the provider side rand context in
+the I<mctx> parameter.
+If it receives NULL as I<ctx> value, it should not do anything other than
+return.
+This function should free any resources associated with that context.
+
+=head2 Random Number Generator Functions: NIST
+
+These functions correspond to those defined in NIST SP 800-90A and SP 800-90C.
+
+OP_rand_instantiate() is used to instantiate the DRBG I<ctx> at a requested
+security I<strength>.  In addition, I<prediction_resistance> can be requested.
+Additional input I<addin> of length I<addin_len> bytes can optionally
+be provided.
+
+OP_rand_uninstantiate() is used to uninstantiate the DRBG I<ctx>.  After being
+uninstantiated, a DRBG is unable to produce output until it is instantiated
+anew.
+
+OP_rand_generate() is used to generate random bytes from the DRBG I<ctx>.
+It will generate I<outlen> bytes placing them into the buffer pointed to by
+I<out>.  The generated bytes will meet the specified security I<strength> and,
+if I<prediction_resistance> is true, the bytes will be produced after reseeding
+from a live entropy source.  Additional input I<addin> of length I<addin_len>
+bytes can optionally be provided.
+
+=head2 Random Number Generator Functions: Additional
+
+OP_rand_nonce() is used to generate a nonce of the given I<strength> with a
+length from I<min_noncelen> to I<max_noncelen>. If the output buffer I<out> is
+NULL, the length of the nonce should be returned.
+
+OP_rand_set_callbacks() is used to supply custom entropy and nonce callbacks.
+Instead of gathering seed material from its usual sources, the DRBG I<ctx>
+should call these functions.
+The I<get_entropy> and I<cleanup_entropy> call backs obtain and release bytes
+of entropy.
+The I<get_nonce> and I<cleanup_nonce> functions obtain and release nonce bytes.
+In all cases, the additional argument I<arg> is passed to the callbacks.
+
+OP_rand_verify_zeroization() is used to determine the of the internal DRBG
+state is zeroed.  This capability is mandated by NIST as part of the self
+tests, it is unlikely to be useful in other circumstances.
```

Comment:
I'm not sure why we need to expose this. If its run as part of the self-tests then its run provider side - so why do we need to expose it in the provider <-> libcrypto interface?

Reply:
It less nasty this way.

The FIPS KATs are required to use the normal API for accessing the algorithms.  I'm using EVP_RAND calls.  This means the KATs deal with EVP_RAND_CTX objects.  They could be able to jump through the _data_ void pointer to the underlying PROV_RAND structure but it's pretty evil.  Moreover, I've changed how the DRBG specific state is stored.  It's a void pointer to secure memory rather than being inside a union in the PROV_RAND structure.  So in addition to jumping the provider boundary, the tests would also need to know the appropriate structure as well.  Not impossible but nasty.

The DRBG unit test also has the same problems.

I'm willing to switch it but don't like the idea.

Reply:
Ok - leave it.